### PR TITLE
Add python bindings for new dataframe APIs

### PR DESCRIPTION
**Pull request by jleibs**
### What
- First pass at implementing APIs for: https://github.com/rerun-io/rerun/issues/7455
- Introduces a new mechanism for directly exposing rust types into the python bridge via a .pyi definition

Example notebook for testing
```
pixi run py-build-examples
pixi run -e examples jupyter notebook tests/python/dataframe/examples.ipynb
```

### Future work:
- More docs / help strings
- Remaining API features

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7357?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7357?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7357)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: rerun_py/rerun_bindings/rerun_bindings.pyi [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @staticmethod
+    def row_id() -> ControlColumnSelector: ...
+
+class TimeColumnDescriptor:
+    """A column containing the time values for when the component data was updated."""
+
+class TimeColumnSelector:
+    """A selector for a time column."""
+
+    def __init__(self, timeline: str): ...
```
Comment (teh-cmc) [inline rerun_py/rerun_bindings/rerun_bindings.pyi:22]: If we still plan to rename this and related types `IndexColumnSomethingSomething`, I guess now would be a better time than later?

File: rerun_py/rerun_bindings/rerun_bindings.pyi [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    def with_dictionary_encoding(self) -> ComponentColumnDescriptor: ...
+
+class ComponentColumnSelector:
+    """A selector for a component column."""
+
+    def __new__(cls, entity_path: str, component_type: ComponentLike): ...
+    def with_dictionary_encoding(self) -> ComponentColumnSelector: ...
+
+class Schema:
+    """The schema representing all columns in a [`Dataset`][]."""
```
Comment (teh-cmc) [inline rerun_py/rerun_bindings/rerun_bindings.pyi:36]:
```suggestion
    """The schema representing all columns in a [`Recording`][]."""

File: rerun_py/rerun_bindings/rerun_bindings.pyi [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    """An archive loaded from an RRD, typically containing 1 or more recordings or blueprints."""
+
+    def num_recordings(self) -> int: ...
+    def all_recordings(self) -> list[Recording]: ...
+
+def load_recording(filename: str) -> Recording:
+    """
+    Load a single recording from an RRD.
+
+    Will raise a `ValueError` if the file does not contain a single recording.
```
Comment (teh-cmc) [inline rerun_py/rerun_bindings/rerun_bindings.pyi:67]: That sounds ambiguous to me:
```suggestion
    Will raise a `ValueError` if the file does not contain exactly one recording.

File: rerun_py/src/dataframe.rs [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+                "ComponentLike input must be a string or Component class.",
+            ));
+        }
+    }
+}
+
+#[pyclass(frozen, name = "Schema")]
+#[derive(Clone)]
+pub struct PySchema {
+    // TODO(jleibs): This gets replaced with the new schema object
```
Comment (teh-cmc) [inline rerun_py/src/dataframe.rs:263]: Guess it doesn't after all?

File: rerun_py/src/dataframe.rs [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+            }
+            None
+        })
+    }
+}
+
+#[pyclass(name = "Recording")]
+pub struct PyRecording {
+    store: ChunkStore,
+    cache: re_dataframe2::external::re_query::Caches,
```
Comment (teh-cmc) [inline rerun_py/src/dataframe.rs:329]: For a more publicly-stable version of this:
```suggestion
    cache: re_dataframe2::QueryCache,

File: rerun_py/src/dataframe.rs [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+                    &mut engine
+                        .iter_entity_paths(&path_filter)
+                        .map(|p| (p, Some(components.clone())))
+                        .collect(),
+                );
+            }
+
+            Ok(contents)
+        } else {
+            return Err(PyTypeError::new_err("ViewContentsLike input must be..."));
```
Comment (teh-cmc) [inline rerun_py/src/dataframe.rs:520]: Is there some pythonic magic going on here or was the err message never finished? :thinking:

File: rerun_py/src/dataframe.rs [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+        contents
+    ))]
+    fn view(
+        slf: Bound<'_, Self>,
+        timeline: &str,
+        contents: Bound<'_, PyAny>,
+    ) -> PyResult<PyRecordingView> {
+        let borrowed_self = slf.borrow();
+
+        // Look up the type of the timelin
```
Comment (teh-cmc) [inline rerun_py/src/dataframe.rs:545]: Staying ahead of `typos`...
```suggestion
        // Look up the type of the timeline

**Review by teh-cmc — APPROVED** [review]

File: rerun_py/src/dataframe.rs (hunk 6 above)
Comment (jleibs) [reply rerun_py/src/dataframe.rs:520]: No magic -- just forgot to finish while I was figuring out what to write.

**Comment by emilk** [comment]
This PR added a dependency on `TypeAlias` which was added in Python 3.10, making Rerun incompatible with earlier Rerun versions:

```
❯ python --version
Python 3.9.12

❯ python test.py /tmp/dna.rrd                                           
Traceback (most recent call last):
  File "/Users/emilk/code/rerun/test.py", line 2, in <module>
    import rerun as rr
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/rerun_sdk/rerun/__init__.py", line 18, in <module>
    import rerun_bindings as bindings  # type: ignore[attr-defined]
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/rerun_bindings/__init__.py", line 3, in <module>
    from typing import TYPE_CHECKING, TypeAlias, Union
ImportError: cannot import name 'TypeAlias' from 'typing' (/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/typing.py)
```

File: rerun_py/rerun_bindings/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, TypeAlias, Union
```
Comment (emilk) [inline rerun_py/rerun_bindings/__init__.py:3]: Why add these imports here?

I removed them in https://github.com/rerun-io/rerun/pull/7794

**Pull request merged**